runtime/include/json.h

```diff
@@ -0,0 +1,284 @@
+		return false;
+	}
+
+	return true;
+}
+
+static inline bool
+has_valid_type(jsmntok_t tok, char *key, jsmntype_t expected_type)
+{
+	if (tok.type != expected_type) {
+		fprintf(stderr, "The value of the key %s should be a %s, was a %s\n", key, jsmn_type(expected_type),
+		        jsmn_type(tok.type));
+		return false;
+	}
+
+	return true;
+}
+
+static inline bool
+is_nonempty_string(jsmntok_t tok, char *key)
+{
+	if (!has_valid_type(tok, key, JSMN_STRING)) return false;
+
+	if (tok.end - tok.start < 1) {
+		fprintf(stderr, "The value of the key %s was an empty string\n", key);
+		return false;
+	}
+
+	return true;
+}
+
+static inline bool
+is_valid_key(jsmntok_t tok)
+{
+	if (tok.type != JSMN_STRING) {
+		fprintf(stderr, "Expected token to be a key with a type of string, was a %s\n", jsmn_type(tok.type));
+		return false;
+	}
+
+	if (tok.end - tok.start < 1) {
+		fprintf(stderr, "Key was an empty string\n");
+		return false;
+	}
+
+	return true;
+}
+
+static inline int
+parse_uint8_t(jsmntok_t tok, const char *json_buf, const char *key, uint8_t *ret)
+{
+	char    *end  = NULL;
+	intmax_t temp = strtoimax(&json_buf[tok.start], &end, 10);
+
+	if (end != &json_buf[tok.end] || temp < 0 || temp > UINT8_MAX) {
+		fprintf(stderr, "Unable to parse uint8_t for key %s\n", key);
+		return -1;
+	}
+
+	*ret = (uint8_t)temp;
+	return 0;
+}
+
+static inline int
+parse_uint16_t(jsmntok_t tok, const char *json_buf, const char *key, uint16_t *ret)
+{
+	char    *end  = NULL;
+	intmax_t temp = strtoimax(&json_buf[tok.start], &end, 10);
+
+	if (end != &json_buf[tok.end] || temp < 0 || temp > UINT16_MAX) {
+		fprintf(stderr, "Unable to parse uint16_t for key %s\n", key);
+		return -1;
+	}
+
+	*ret = (uint16_t)temp;
+	return 0;
+}
+
+static inline int
+parse_uint32_t(jsmntok_t tok, const char *json_buf, const char *key, uint32_t *ret)
+{
+	char     *end  = NULL;
+	uintmax_t temp = strtoimax(&json_buf[tok.start], &end, 10);
+
+	if (end != &json_buf[tok.end] || temp < 0 || temp > UINT32_MAX) {
+		fprintf(stderr, "Unable to parse uint32_t for key %s\n", key);
+		return -1;
+	}
+
+	*ret = (uint32_t)temp;
+	return 0;
+}
+
+/**
+ * Parses a JSON file into an array of module configs
+ * @param file_name The path of the JSON file
+ * @return module_config_vec_len on success. -1 on Error
+ */
+static inline int
+parse_json(const char *json_buf, ssize_t json_buf_size, struct module_config **module_config_vec)
+{
+	assert(json_buf != NULL);
+	assert(json_buf_size > 0);
+	assert(module_config_vec != NULL);
+
+	jsmntok_t tokens[JSON_MAX_ELEMENT_SIZE * JSON_MAX_ELEMENT_COUNT];
+	int       module_config_vec_len = 0;
+
+
+	/* Initialize the Jasmine Parser and an array to hold the tokens */
+	jsmn_parser module_parser;
+	jsmn_init(&module_parser);
+
+	/* Use Jasmine to parse the JSON */
+	int total_tokens = jsmn_parse(&module_parser, json_buf, json_buf_size, tokens,
+	                              sizeof(tokens) / sizeof(tokens[0]));
```

Comment:
Why not use `JSON_MAX_ELEMENT_SIZE * JSON_MAX_ELEMENT_COUNT` again instead of `sizeof(tokens) / sizeof(tokens[0])` ?

Reply:
Actually why not just declare a single constant (e.g. JSON_MAX_SIZE=16*1024) that is equal to the product of JSON_MAX_ELEMENT_SIZE * JSON_MAX_ELEMENT_COUNT ? Since you are using them only once and multiplying anyways.